File: packages/server/src/plugins/channel.ts
import type { FastifyReply } from "fastify"
import type { Logger } from "../logger"

export interface PluginOutboundEvent {
  type: string
  properties?: Record<string, unknown>
}

interface ClientConnection {
  reply: FastifyReply
  workspaceId: string
}

export class PluginChannelManager {
  private readonly clients = new Set<ClientConnection>()

  constructor(private readonly logger: Logger) {}

  register(workspaceId: string, reply: FastifyReply) {
    const connection: ClientConnection = { workspaceId, reply }
    this.clients.add(connection)
    this.logger.debug({ workspaceId }, "Plugin SSE client connected")

    let closed = false
    const close = () => {
      if (closed) return
      closed = true
      this.clients.delete(connection)
      this.logger.debug({ workspaceId }, "Plugin SSE client disconnected")
    }

    return { close }
  }

  send(workspaceId: string, event: PluginOutboundEvent) {
    for (const client of this.clients) {
      if (client.workspaceId !== workspaceId) continue
      this.write(client.reply, event)
    }
  }

  broadcast(event: PluginOutboundEvent) {
    for (const client of this.clients) {
      this.write(client.reply, event)
    }
  }

  private write(reply: FastifyReply, event: PluginOutboundEvent) {
    try {
      reply.raw.write(`data: ${JSON.stringify(event)}\n\n`)
    } catch (error) {
      this.logger.warn({ err: error }, "Failed to write plugin SSE event")
    }
  }
}

File: packages/server-rust/src/plugins/channel.rs
use std::collections::{HashSet, HashMap};
use tokio::sync::broadcast;

use crate::logger::Logger;

#[derive(Debug, Clone)]
pub struct PluginOutboundEvent {
    pub event_type: String,
    pub properties: Option<HashMap<String, serde_json::Value>>,
}

pub struct PluginChannelManager {
    clients: HashSet<String>,
    tx: broadcast::Sender<serde_json::Value>,
    logger: Logger,
}

impl PluginChannelManager {
    pub fn new(logger: Logger) -> Self {
        let (tx, _) = broadcast::channel(256);
        Self {
            clients: HashSet::new(),
            tx,
            logger,
        }
    }

    pub fn register(&mut self, workspace_id: &str) {
        self.clients.insert(workspace_id.to_string());
        tracing::debug!(component = %self.logger.component, workspace_id = %workspace_id, "Plugin SSE client connected");
    }

    pub fn unregister(&mut self, workspace_id: &str) {
        self.clients.remove(workspace_id);
        tracing::debug!(component = %self.logger.component, workspace_id = %workspace_id, "Plugin SSE client disconnected");
    }

    pub fn send(&self, workspace_id: &str, event: &PluginOutboundEvent) {
        if self.clients.contains(workspace_id) {
            let payload = serde_json::json!({
                "workspaceId": workspace_id,
                "event": event.event_type,
                "properties": event.properties,
            });
            let _ = self.tx.send(payload);
            tracing::debug!(
                component = %self.logger.component,
                workspace_id = %workspace_id,
                event_type = %event.event_type,
                "Sending plugin event"
            );
        }
    }

    pub fn broadcast(&self, event: &PluginOutboundEvent) {
        if !self.clients.is_empty() {
            let payload = serde_json::json!({
                "event": event.event_type,
                "properties": event.properties,
            });
            let _ = self.tx.send(payload);
            tracing::debug!(
                component = %self.logger.component,
                client_count = self.clients.len(),
                event_type = %event.event_type,
                "Broadcasting plugin event"
            );
        }
    }

    pub fn subscribe(&self) -> broadcast::Receiver<serde_json::Value> {
        self.tx.subscribe()
    }

    pub fn publish(&mut self, value: serde_json::Value) -> usize {
        self.tx.send(value).unwrap_or(0)
    }

    pub fn sender(&self) -> broadcast::Sender<serde_json::Value> {
        self.tx.clone()
    }
}
